Lib/test/test_logging.py

```diff
@@ -4310,6 +4311,20 @@ def test_issue_89047(self):
             s = f.format(r)
             self.assertNotIn('.1000', s)
 
+    def test_issue_102402_100msecs(self):
+        tests = (
+            # (time_ns, expected_msecs_value)
+            (1_677_902_297_100_000_000, 100.0),  # exactly 100ms
+            (1_677_903_920_999_998_503, 999.0),  # check truncating doesn't round
+            (1_677_903_920_000_998_503, 0.0),  # check truncating doesn't round
+
+        )
+        for ns, want in tests:
+            with patch('time.time_ns') as patched_ns:
+                patched_ns.return_value = ns
+                record = logging.makeLogRecord({'msg': 'test'})
+            self.assertEqual(record.msecs, want)
```

Comment:
Test also `created` and `relativeCreated` using `assertAlmostEqual()`. It is needed to ensure that they leave approximately the same, but `relativeCreated` should be now more accurate, and this can be detected if choose an appropriate `_startTime`.

Reply:
Hmm... I'm not sure I understand. Can you elaborate?

This is what I understood:

```python
for ns, want in tests:
    ...
    self.assertEqual(record.msecs, want)
    self.assertAlmostEqual(record.created, record.relativeCreated)
```

However, that won't work because `relativeCreated` is set on module load, before we patch `time_ns()`, and they represent two vastly different things. I guess if we assume that `_startTime = 0`, then `self.assertAlmostEqual(record.created, record.relativeCreated)` would be `True`.

Were you thinking to also patch `logging._startTime` to a known value? Perhaps something like:

```python
    def test_issue_102402_relativeCreated(self):
        mok_startTime = 1_677_903_920_000_998_503
        delta_ns = 30_000_000
        ns = mok_startTime + delta_ns
        with patch.object(logging, "_startTime", mok_startTime):
            with patch("time.time_ns") as patched_ns:
                patched_ns.return_value = ns
                record = logging.makeLogRecord({'msg': 'test'})
        self.assertAlmostEqual(record.created, ns / 1e9)
        self.assertAlmostEqual(record.relativeCreated, 30)
```

Though I'm not sure how useful of a test the above example actually is... It seems like it's mostly testing _itself_.

Hence, I ask for more information 😁.

Reply:
Yes, you need also to patch `_startTime`. The test is needed to check that there is no some silly error in computing `relativeCreated`, like using different scale. But we can also test that it is more precise now. Note that `assertAlmostEqual` allows you to specify `places` and `delta`, so you can test more precisely that by default.

Thinking about this more, it is better to not set `_startTime` explicitly, but patch `time.time_ns`, import fresh `logging` module, change the return value of the patched `time.time_ns`, and then make the record.

Reply:
I think I understand now. PTAL at the new test. Ran the new test on `main` (well, cb944d0be8 as this PR is so old haha) and found that `relativeCreated` was only able to get about 3 places of precision. With this PR, that bumped to 7 places.